doc/decisions/0_drafts/errors_core.md

```diff
@@ -278,12 +287,77 @@ if (found == NULL)
 
 ## Decision
 
-**Suggestion:** Go with "Pragmatic solution"
+Go with "Pragmatic solution".
+We don't define any strict rules for the entire API.
+Instead, we define a few basic guidelines.
+
+### Guidelines
+
+These guidelines shall be applied to each function of the public API separately to decide how the function treats errors.
+
+#### Boolean functions
+
+If a function is conceptually boolean, it should return a `bool`.
+This means choosing `true` or `false` as the fallback value used in case of errors.
+
+If choosing `true` or `false` is not feasible, redesigning the function may be necessary.
+The function may be doing to many checks in one go and splitting it into multiple functions could make sense.
+
+If splitting is also not possible, an `int` or `enum` should be returned.
+The name of such a function must make clear that it is not actually boolean (avoid `is` and `has`, prefer `check`) to avoid accidental use in an `if`.
+
+#### Check preconditions
+
+Functions must check all preconditions.
+A (bug-free) function should never segfault, abort the process or otherwise cause unexpected behavior on the caller side.
+All conditions that could cause such behavior must be listed as preconditions and checked by the function as a safety measure.
+
+#### `NULL` arguments
+
+A `NULL` pointer argument should never be treated as a separate error case (unless it is the only error case).
+If a `NULL` pointer constitutes an invalid argument, the function should return an existing error (might be used for other precondition checks) or a default fallback value.
+
+#### Simple preconditions
+
+Simple preconditions are those that can be checked easily by the caller and which are not computationally expensive to check.
+All such preconditions should use the same error code, which may even be shared with other error cases.
+
+#### Complex preconditions
+
+Complex preconditions which
+
+- cannot (easily) be checked by the caller (e.g., because there is no public API for it) OR
+- are computationally expensive to check (i.e., duplicating the check would be bad)
+
+should always use separate error codes for each error case.
+The error code may not be shared with other complex preconditions, but can be shared with simple preconditions or `NULL` arguments, since those are easy to distinguish for the caller.
+
+#### Restructure API
+
+To avoid complex preconditions and the need for multiple error codes, the API can be restructured.
+This should, however, be done with caution.
+The API should still remain ergonomic to use and there should not be extra memory allocations creating opaque objects just to guard against possible errors.
+
+#### Error arguments
+
+If a function needs multiple error codes, return them as an `int` or enum value.
+Do not use an error argument (like `Key * error`).
+Instead, use an output pointer argument (e.g., `Foo * output`) for the data produced by the function.
+This makes it much more obvious to the caller that errors can occur.
```

Comment:
Rules must be more specific. And as you say yourself in the next paragraph, this cannot be a general rule, as it is library-depend. Please let us focus on core in this decision, otherwise we won't reach anywhere.

Reply:
> Rules must be more specific.

Which is why I called them guidelines and even explicitly stated that we don't define strict rules.

A rule should cover all cases and should be followed unless there is a specific exception.

A guideline is just a guide. You should follow it as much as possible, but it is very much expected that you may have to deviate a bit.

> And as you say yourself in the next paragraph, this cannot be a general rule, as it is library-depend. Please let us focus on core in this decision, otherwise we won't reach anywhere.

This is actually already pretty concrete. But if you want I can make it **very** concrete and also library-independent:

---

If a function needs multiple error codes, prefer returning them as an `int` or enum value.

```c
// returns 0 on success
// returns -1 if key == NULL, or a precondition is violated
// returns -2 if error X happens
int foo (Key * key);
```

> **Note:** Sorry, it's late I couldn't come up with a better example, but I think this first part is pretty clear even without a concrete example.

Error arguments should only be used, if the errors are likely to be shown to a user.
Only `Key * errorKey` or a struct which follows the same error concept (e.g. `ElektraError` from the high-level API) should be used as an error argument.
The `Key * errorKey` concept was explicitly designed with the intention of having errors that can be shown to users, therefore it should be used as designed.

```c
/*
 * @retval 0	on success, this is the only case in which @p ks will be modified
 * @retval -1	on error, the error will be set as metadata in @p errorKey
 * @retval 1	if the help option `--help` was found, use elektraGetOptsHelpMessage() access the
 * 		generated help message
 */
int elektraGetOpts (KeySet * ks, int argc, const char ** argv, const char ** envp, Key * parentKey);

/**
 * @param error Pointer used to report errors
 * @retval True if the checks were successful (or not required).
 * @retval False if the checks were required but unsuccessful.
 */
kdb_boolean_t checkSpec (Key * const parentKey, KeySet * contract, ElektraError ** error)
```

If the function produces data, but also needs multiple error codes, prefer returning the error code and using output pointer argument(s) for the data.
By not returning a value other than the data the caller asked for, it is more clear that the value may convey error information.

```c
/*
 * @param variable pointer to the output variable
 * @retval 0 on success
 * @retval -1 on precondition violation (key == NULL, variable == NULL)
 * @retval -2 if key has the wrong meta:/type
 * @retval -3 if the value of key could not be converted
 */
int elektraKeyToChar (const Key * key, kdb_char_t * variable);
```
---